packages/dds/merge-tree/docs/Attribution.md

```diff
@@ -0,0 +1,378 @@
+
+```typescript
+export interface IAttributor {
+  getAttributionInfo(seq: number): AttributionInfo;
+}
+
+export const binByMinutes = (interval: number) => (timestamp: number) => timestamp % (1000 * 60 * interval);
+
+const seqComparator = (a: AttributionEntry, b: AttributionEntry) => {
+  aEnd = typeof a.k === 'number' ? a.k : a.k[1];
+  bEnd = typeof a.k === 'number' ? b.k : b.k[1];
+  return aEnd - bEnd;
+}
+
+class Attributor implements IAttributor {
+  private seqToInfo: SortedList<AttributionEntry> = new SortedList(seqComparator);
+  constructor(
+    runtime: IFluidDataStoreRuntime,
+    serialized?: SerializedAttributor,
+    bin: (timestamp: number) => number = binByMinutes(5)
+  ) {
+    if (serialized) {
+      const interner = new ObjectInterner(serialized.interner);
+      // Note: this implementation doesn't coalesce re-binned attribution entries that are newly equivalent and adjacent.
+      this.seqToInfo.extend(...serialized.lookup.map(({ k, v: internedV }) => {
+        const { timestamp, userRef } = isInternedRef(maybeInternedV) ? interner.getObject(maybeInternedV) : maybeInternedV;
+        const v = {
+          timestamp: bin(timestamp),
+          user: interner.getObject(userRef)
+        };
+        return { k, v };
+      }));
+    }
+
+    const { deltaManager, audience } = runtime;
+    deltaManager.on("op", (message: ISequencedDocumentMessage) => {
+      const attributionInfo = {
+         /* note: for object interning to work, this needs to be a referentially equal user object. If that isn't provided by the
+            fluid framework, we probably would want a layer of caching here. For interning of overall attribution info objects,
+            we may want a similar cache. */
+        user: audience.get(message.clientId).user,
+        timestamp: bin(message.timestamp)
+      };
+      const { k, v } = seqToInfo.getAt(seqToInfo.length - 1);
+      const lastEntryStart = typeof k === 'number' ? k : k[0];
+      const lastEntryEnd = typeof k === 'number' ? k : k[1];
+      if (
+        attributionInfosAreEquivalent(attributionInfo, v) &&
+        // Note: this coalescing logic is somewhat unideal since no-ops break it.
+        message.seq === 1 + lastEntryEnd)
+      ) {
+        this.seqToInfo.pop();
+        this.seqToInfo.insert({ k: [lastEntryStart, message.seq], v });
+      } else {
+        this.seqToInfo.insert({ k: message.seq, v: attributionInfo });
+      }
+    });
+  }
+
+  public getAttributionInfo(seq: number): AttributionInfo {
+    const { k, v } = seqToInfo.findAtOrAfter(seq);
+    assert(k === seq || (k.length === 2 && k[0] <= seq && seq <= k[1]));
+    return v;
+  }
+
+  // Unpictured:
+  // - serialization (not interesting; deserialization logic is pictured)
+  // - GC (there are several ways to hook this up, though one can check the data structure should support it in O(n))
+}
+
+```
+
+
+### Bookkeeping Placement Considerations
```

Comment:
even given my comments above, i'm still pretty open to where we land here, but it think considering storing the op info (seq, timestamp, op,...) compressed and near the dds, and only keeping user info in the upper layers could be a good compromise.